mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -0,0 +1,104 @@
+//===- ArithToEmitC.cpp - Arith to EmitC conversion -----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to convert arith ops into emitc ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/ArithToEmitC/ArithToEmitC.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_ARITHTOEMITCCONVERSIONPASS
+#include "mlir/Conversion/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+
+static bool isConvertibleToEmitC(Type type) {
```

Comment:
This is a rather broad check which could also be reused in other conversions. Furthermore, it mainly models a restriction which is introduced by the emitter but not the dialect itself, where both can change. Thus I think this check at least needs to be move to a more accessible location to be reused by other conversions and ideally verifiers or even the emitter. And validation pass could be another alternative, at least until the EmitC dialect and the emitter match more closely.

Reply:
Yes, I agreed the check should be used more widely. I propose we do the following:
1) We restrict the types in the EmitC dialect (tablegen) by a function like this, which aligns with what the emitter currently supports. Later on, we could consider to restrict the types further if it makes sense. I can create a separate PR for this.
2) We drop the check in the conversion for now, and after 1) lands, we enable it again in the conversion such that the conversion rejects invalid types rather than generating invalid EmitC.
What do you think?

Reply:
This is exactly what I have proposed offline to @simon-camp. So yeah, sounds good to me :+1: